apps/render_cf.py

```diff
@@ -11,35 +12,245 @@ def snake_to_pascal_case(input_string: str):
+
+
+def get_batch_param_names_for_job_step(step: dict) -> set[str]:
+    ref_prefix = 'Ref::'
+    return {
+        arg.removeprefix(ref_prefix)
+        for arg in step['command']
+        if arg.startswith(ref_prefix)
+    }
+
+
+def render_templates(job_types, compute_envs, security_environment, api_name):
+    job_states = get_states_for_jobs(job_types)
+
     env = jinja2.Environment(
         loader=jinja2.FileSystemLoader('./'),
         autoescape=jinja2.select_autoescape(default=True, disabled_extensions=('j2',)),
         undefined=jinja2.StrictUndefined,
         trim_blocks=True,
         lstrip_blocks=True,
         keep_trailing_newline=True,
+        extensions=['jinja2.ext.do'],
     )
 
     for template_file in Path('.').glob('**/*.j2'):
         template = env.get_template(str(template_file))
 
         output = template.render(
             job_types=job_types,
+            compute_envs=compute_envs,
+            compute_env_names=[env['name'] for env in compute_envs],
             security_environment=security_environment,
             api_name=api_name,
             json=json,
             snake_to_pascal_case=snake_to_pascal_case,
+            job_states=job_states,
         )
 
+        if str(template_file).endswith('.json.j2'):
+            output = json.dumps(json.loads(output), indent=2)
+
         template_file.with_suffix('').write_text(output)
 
 
+def get_compute_environments(job_types: dict, compute_env_file: Optional[Path]) -> list[dict]:
+    compute_envs = []
+    compute_env_names = set()
+    compute_env_imports = set()
+    for _, job_spec in job_types.items():
+        for step in job_spec['steps']:
+            compute_env = step['compute_environment']
+            if 'name' in compute_env:
+                name = compute_env['name']
+                assert name != 'Default'
+                if name in compute_env_names:
+                    raise ValueError(
+                        f'Compute envs must have unique names but the following is defined more than once: {name}.'
+                    )
+                compute_envs.append(compute_env)
+                compute_env_names.add(name)
+            elif 'import' in compute_env and compute_env['import'] != 'Default':
+                compute_env_imports.add(compute_env['import'])
+            else:
+                assert compute_env['import'] == 'Default'
+
+    if compute_env_file:
+        compute_envs_from_file = yaml.safe_load(compute_env_file.read_text())['compute_environments']
+        for name in compute_envs_from_file:
+            assert name != 'Default'
+            if name in compute_env_names:
+                raise ValueError(
+                    f'Compute envs must have unique names but the following is defined more than once: {name}.'
+                )
+            compute_env = compute_envs_from_file[name]
+            compute_env['name'] = name
+            compute_envs.append(compute_env)
+            compute_env_names.add(name)
```

Comment:
This results in all of the compute envs defined in `job_spec/config/compute_environments.yml` getting added to the `compute_envs` list, so they get created in all deployments, even if those compute envs were not actually imported in any of the job specs in the `job_types` list.

Reply:
We may also want to add unit tests for this function (and probably some of the other render functions too).